File: Search2DMatrixII.cpp
// Time Complexity :  O(n + m) where n, m = input array size. 
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes

// Three line explanation of solution in plain english

/* Variable i -> row index and j -> column index of pointer.
 * Assign a pointer to top right corner of matrix. Repeat the following process until the pointer is within the bounds of matrix dimensions.
 *     If matrix[i][j] is equal to target then return true. If matrix[i][j] > target then decrement the column index j else increment row index i.
 * If loop terminated then return false indicating element not present in matrix.
 * 
 * Same thing can also be done by starting from bottom left of array.
 */

class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        if (matrix.size() == 0)
            return false;
        int i = 0;
        int j = matrix[0].size() - 1;
        
        while (i < matrix.size() && j >= 0)
        {
            if (matrix[i][j] == target)
                return true;
            
            if (target < matrix[i][j])
                j--;
            else
                i++;
        }
        return false;
    }
};

File: search2DmatrixII.py
class Solution:
    def searchMatrix(self, matrix: list[list[int]], target: int) -> bool:
        if matrix == None or len(matrix)==0:
            return False
        
        m=len(matrix)
        n=len(matrix[0])

        #starting from upper right cornor
        # row = 0
        # col = n-1
        # while col >=0 and row < m:
        #     if target == matrix[row][col]:
        #         return True
        #     elif target > matrix[row][col]:
        #         row+=1
        #     else:
        #         col-=1
        # return False


        #starting from lower left cornor
        col=0
        row=m-1
        while row>=0 and col<n:
            if target == matrix[row][col]:
                return True
            elif target > matrix[row][col]:
                col+=1
            else:
                row-=1
        return False
    
    
#time complexity -> O(m+n)
#space complexity -> O(1)